File: src/main/scala/com/tokoko/spark/flight/FlightManager.java
package com.tokoko.spark.flight;

import com.google.protobuf.ByteString;
import org.apache.arrow.flight.*;
import org.apache.arrow.memory.BufferAllocator;

import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.stream.Collectors;

public class FlightManager {

    private final HashMap<ByteString, String> flights;

    private final BufferAllocator allocator;

    private final List<FlightClient> clients;

    public FlightManager(List<Location> locations, BufferAllocator allocator) {
        this.allocator = allocator;
        flights = new HashMap<>();

        clients = locations.stream()
                .map(loc -> Location.forGrpcInsecure(loc.getUri().getHost(), loc.getUri().getPort()))
                .map(location -> FlightClient.builder(allocator, location).build())
                .collect(Collectors.toList());
    }

    public void addFlight(ByteString handle) {
        if (flights.containsKey(handle)) {
            throw CallStatus.ALREADY_EXISTS.toRuntimeException();
        }
        flights.put(handle, "RUNNING");
    }

    public void setCompleted(ByteString handle) {
        if (!flights.containsKey(handle)) {
            throw new RuntimeException("Flight doesn't exist");
        }
        flights.put(handle, "COMPLETED");
    }

    public String getStatus(ByteString handle) {
        if (!flights.containsKey(handle)) {
            throw new RuntimeException("Flight doesn't exist");
        }
        return flights.get(handle);
    }

    public void broadcast(ByteString handle) {
        if (!flights.containsKey(handle)) {
            throw new RuntimeException("Flight doesn't exist");
        }

        Action action = new Action(flights.get(handle), handle.toByteArray());

        clients.forEach(client -> {
                Iterator<Result> it = client.doAction(action);

                while (it.hasNext()) {
                    System.out.println(it.next().getBody().length);
                }
        } );
    }

}

File: src/main/scala/com/tokoko/spark/flight/FlightManager.scala
package com.tokoko.spark.flight

import com.google.protobuf.ByteString
import org.apache.arrow.flight._
import org.apache.arrow.memory.BufferAllocator
import java.util

class FlightManager(val locations: List[Location], val allocator: BufferAllocator) {

  private val flights = new util.HashMap[ByteString, String]

  private val clients = locations
    .map((loc: Location) => Location.forGrpcInsecure(loc.getUri.getHost, loc.getUri.getPort))
    .map((location: Location) => FlightClient.builder(allocator, location).build)


  def addFlight(handle: ByteString): Unit = {
    if (flights.containsKey(handle)) throw CallStatus.ALREADY_EXISTS.toRuntimeException
    flights.put(handle, "RUNNING")
  }

  def setCompleted(handle: ByteString): Unit = {
    if (!flights.containsKey(handle)) throw new RuntimeException("Flight doesn't exist")
    flights.put(handle, "COMPLETED")
  }

  def getStatus(handle: ByteString): String = {
    if (!flights.containsKey(handle)) throw new RuntimeException("Flight doesn't exist")
    flights.get(handle)
  }

  def broadcast(handle: ByteString): Unit = {
    if (!flights.containsKey(handle)) throw new RuntimeException("Flight doesn't exist")
    val action = new Action(flights.get(handle), handle.toByteArray)

    clients.foreach((client: FlightClient) => {
      val it = client.doAction(action)
      while (it.hasNext) println(it.next.getBody.length)
    })
  }
}
